jpeg/jpeg.go

```diff
@@ -0,0 +1,91 @@
+// jpeg handles scrubbing jpeg files of metadata
+package jpeg
+
+import (
+	"bytes"
+	"encoding/binary"
+	"io"
+	"io/ioutil"
+
+	jpegimagestructure "github.com/dsoprea/go-jpeg-image-structure"
+)
+
+var headerBytes []byte = []byte{0xff, 0xd8}
+
+type JpegScrubber struct {
+	reader      io.Reader
+	segmentData io.LimitedReader
+	wroteHeader bool
+}
+
+func NewJpegScrubber(reader io.Reader) *JpegScrubber {
+	return &JpegScrubber{
+		reader: reader,
+	}
+}
+
+func isMetadataType(segmentMarker []byte) bool {
+	// TODO: consider other APP types
+	return segmentMarker[0] == 0xff &&
+		(segmentMarker[1] == jpegimagestructure.MARKER_APP1 || segmentMarker[1] == jpegimagestructure.MARKER_APP0)
+}
+
+func (self *JpegScrubber) Read(buf []byte) (n int, err error) {
+	var n2 int
+
+	if !self.wroteHeader {
+		n, err = io.ReadFull(self.reader, buf[:len(headerBytes)])
+		if err != nil {
+			return
+		}
+		self.wroteHeader = true
+	}
+
+	n2, err = self.segmentData.Read(buf[n:])
+	n += n2
+	if err != io.EOF {
```

Comment:
I think you want to check for non-nil and non-EOF errors.  `self.segmentData` reads from `self.reader` which can return who-knows-what errors.

Reply:
Huh, I thought that's what I was doing? Basically the only acceptable state for `err` to be in after line 45 is `io.EOF`. I guess if `err` is `nil` there's a fairly serious logic flaw in my code.... I wonder what the correct way to represent that would be...

Reply:
Whoops, yeah you're right!

Actually, you can get a `nil` error here and you _do_ want to return in that case.  It means that there is still data left to read in the segment, but you've filled the buffer.  TBH, I was thinking the same thing as you, but I put some print statements in, ran the tests, and it made sense after that.

Anyway, sorry for the confusion, this is correct as-is =)